File: android/app/src/main/java/org/rhythmeta/maimaid/core/data/LetterGameMatchPresentationPolicy.kt
package org.rhythmeta.maimaid.core.data

object LetterGameMatchPresentationPolicy {
    fun shouldAccept(status: String, matchId: String, trackedMatchId: String?): Boolean =
        status == "active" || matchId == trackedMatchId

    fun visiblePlayers(
        players: List<LetterGameMatchPlayer>,
        roomMembers: List<LetterGameRoomMember>,
    ): List<LetterGameMatchPlayer> {
        val acceptedUserIds = roomMembers
            .asSequence()
            .filter { it.status == "accepted" }
            .map(LetterGameRoomMember::userId)
            .toSet()
        return players.filter { it.userId in acceptedUserIds }
    }
}

File: ios/maimaid/Models/LetterGameMatchPresentationPolicy.swift
import Foundation

nonisolated enum LetterGameMatchPresentationPolicy {
    static func shouldAccept(status: String, matchId: String, trackedMatchId: String?) -> Bool {
        status == "active" || matchId == trackedMatchId
    }

    static func visiblePlayers(
        _ players: [LetterGameMatchPlayer],
        roomMembers: [LetterGameRoomMember]
    ) -> [LetterGameMatchPlayer] {
        let acceptedUserIds = Set(
            roomMembers.lazy
                .filter { $0.status == "accepted" }
                .map(\.userId)
        )
        return players.filter { acceptedUserIds.contains($0.userId) }
    }
}
